### ISSUE TITLE
Add destroy/deconstructor for Backbone.Events

### ISSUE THREAD

**Issue by nmschulte** [jashkenas/backbone#4039]
Doesn't it make sense to have `Backbone.Events` objects automatically call `off` (with no arguments; to unregister all `once` and `on` handlers) when being destroyed?

Particularly, this is useful for `Backbone.View` objects (or `Backbone.Model` objects, though I haven't run into a use case for this) that listen to themselves (mainly, when using Marionette's views/trigger methods).

This could be done by having the `Backbone.View`'s `destroy` call `off`, but the refactor to give `Backbone.Events` a deconstructor seems a better solution to me.

For now, I've been using `listenTo`/`listenToOnce` instead of using `on`/`once` and managing an `off` call.

**Comment by paulfalgout** [comment]
Backbone.View doesn't have a destroy to do clean up in.  I don't think View "destroy" is a Backbone concern.  

However whenever you `View#remove` in Backbone it does call `stopListening` https://github.com/jashkenas/backbone/blob/master/backbone.js#L1276

Marionette does behave as you're describing, so if you're using Marionette, it is already doing this.  https://github.com/marionettejs/backbone.marionette/blob/a9d9c351d15914d476ac7cffbf65268abc73a6dd/test/unit/item-view.spec.js#L258

Backbone Models also call `stopListening` on `destroy` https://github.com/jashkenas/backbone/blob/master/backbone.js#L673

`stopListening` will work because `this.on` is essentially `this.listenTo(this` 

So I think the cleanup is happening.  But if you have a fiddle or something where you'd expect a listener to get cleaned up but it isn't I'll certainly take a look.  Feel free to ask on https://gitter.im/marionettejs/backbone.marionette as well!

**Comment by nmschulte** [comment]
> Backbone.View doesn't have a destroy to do clean up in. I don't think View "destroy" is a Backbone concern.

Ah yes, apologies: `s/destroy/remove/`; I've been playing w/ Marionette a bit much it seems, :).

To your point about `stopListening`: while the implementation does use `off` (and `listenTo` is similar to `on`), they are not the same.  If you look, `off` is only called for each handler setup w/ `listenTo`/`listenToOnce`; calls to `on` don't track the handler, and therefor will never be cleaned up.

Simply calling `off` w/ no arguments should clean up everything (evidenced by the fact that `stopListening` just maps handlers to `off` calls).

The point that `Backbone.Model` and `Backbone.View` don't have a common destructor (or any of the Backbone modules, for that matter) probably puts a nail in the "add a destructor (destroy/remove/cleanup) to Backbone.Event."  The inheritance philosophy of Backbone is my biggest gripe, which is what leads us here.

`Backbone.Model` should probably just call `off` w/o args, as should `Backbone.View`, and Marionette's test should probably test that `off` is called w/o args.

**Comment by paulfalgout** [comment]
I mean I guess I see what you're saying, but I don't know the purpose of removing listeners on itself.

The thing is `this.on` and subsequently `this.listenTo(this` _don't_ need to be cleaned up because they only reference themselves. Once the view is ready for garbage collection, the listeners will go with it.  If you had the need for some other context then you'd be using `listenTo` correct?

Maybe if there was a more specific example?

Had to dig back a bit, but this was discussed on Marionette for Controller https://github.com/marionettejs/backbone.marionette/issues/1219

**Comment by nmschulte** [comment]
> The thing is this.on and subsequently this.listenTo(this don't need to be cleaned up because they only reference themselves. Once the view is ready for garbage collection, the listeners will go with it. If you had the need for some other context then you'd be using listenTo correct?

I see your point here too.  Is that really how GC works though?  Can it know that this object only references objects that it alone references?  (Does it do the graph traversal?)

**Comment by paulfalgout** [comment]
That is my understanding..  @jmeas or @jridgewell might be able to better answer that?

**Comment by jridgewell** [comment]
>  I guess the on/off issue is moot, so long as I am destroying references to the objects (Backbone.Views) themselves;

Exactly. If you're destroying the view, remove you're references to it.

**Comment by nmschulte** [comment]
I assume Backbone takes the stance that one shouldn't re-use (re-initialize) a `View` that's been `remove`'d?  In that case, I propose renaming `remove` to `destroy`.  If not, I think my point about needing to call `off` still stands (for one, because of the intent of `remove`, and two, to avoid memory leaks): the `Backbone.Events` object will maintain references to the `callback`s, thus preventing them (and their object trees) from being cleaned up.  Even if that is the stance, doesn't it still make sense to call `off` (cleanup non-managed handlers) so that the `callback`s and their reference trees don't get blocked for cleanup during GC until the `Backbone.Events` object is cleaned up?

**Comment by jmeas** [comment]
> I assume Backbone takes the stance that one shouldn't re-use (via e.g. re-initialize) a View that's been remove'd?

I don't think this assumption is right. The docs don't mention anything like this, and [what happens in `remove()`](https://github.com/jashkenas/backbone/blob/9414b9a7c32e3cba1052a42702a86908ba84bc46/backbone.js#L1020-L1026) doesn't do anything that would make it particularly difficult to reuse a view that has been removed.

When I was using Backbone more often, I would often dig through issues to see if I could find opinions from contributors or heavy users of Backbone/Marionette. It's a cool way to see if others agree with your assumptions, or to see a different perspectives on things. I'm not certain, but I wouldn't be surprised if a similar conversation has happened about reusing a view that has been removed.

Anyway, I think that the name `remove` is pretty good, and I wouldn't change it to something like `destroy`, which sounds a bit too heavy for what the method is doing.

> the Backbone.Events object will maintain references to the callbacks

The Backbone.Events object isn't global; it's merged into the View itself. So the view is holding onto its own callbacks. This is perhaps the biggest difference between `listenTo` and `on`, actually; your choice determine which object holds the callbacks and context reference. For this reason, I don't think there's much reason to call `off` within `remove`.

**Comment by nmschulte** [comment]
> I don't think this assumption is right. The docs don't mention anything
like this, and what happens in remove() doesn't do anything that would make
it particularly difficult to reuse a view that has been removed.

> Anyway, I think that the name remove is pretty good, and I wouldn't
change it to something like destroy, which sounds a bit too heavy for what
the method is doing.

I agree 100%.  Someone has to take a stance, preferably the devs.  Anyone
can publish conglomorates of code; Backbone is a toolkit.  If `remove`
doesn't preclude re-use, then it should `off()` (so that re-use can
`on(...)`.

> The Backbone.Events object isn't global;

I know this; s/Backbone.Events/Backbone.Events-derived/.

> Closed #4039.

This is a joke, right?  It's the weekend now; I'll open this again on
Monday.

**Comment by jmeas** [comment]
> If `remove`
> doesn't preclude re-use, then it should `off()` (so that re-use can
> `on(...)`; same reason for calling `stopListening`).

Can you provide a specific example where `.off()` not getting called is causing a problem?

> I know this; s/`Backbone.Events`/`Backbone.Events`-derived/.

Ah, okay, gotcha. You made it out that Backbone.Events could prevent GC; I only meant to explain how it wouldn't for self-references.

> This is a joke, right?  It's the weekend now; I'll open this again on Monday.

I don't think it's a joke! If you can make a compelling argument for this functionality, then there's always a chance for the issue to be reopened :)

**Comment by nmschulte** [comment]
I've thought about this some more, and I see two issues.  The first is w/ behavior, and should be evident from this first example.
If View is constructed/initialized, but never rendered before it is removed/re-used, `Backbone.View` does not what you would expect: multiple handlers are setup for the `'render'`/`_setupEvents` callback.
```
var View = Backbone.View.extend({
    initialize: function () {
        // this won't get cleaned up!
        this.once('render', this._setupEvents);
    },
    _setupEvents: function () {
        this.listenTo(this.options.modelObject, 'change:property', this.render); // heavy-weight update
        this.listenTo(this.options.modelObject, 'change:property2', this._renderProperty2); // light-weight update
    }
});

// --

var obj = ...;
var $el = ...;
var view = new View({modelObject: obj});
beginAsyncStuff(asyncSuccess);

function asyncSuccess() {
    $el.html(view.render().$el);
}

function asyncFailure() {
    var otherObj = ...;
    view.initialize({modelObject: otherObj});
    $el.html(view.render().$el); // e.g.
}
```

The other issue, regarding memory leaks/GC, follows: if handlers are setup on the `Backbone.Events`-derived object with callbacks that are not only referenced by the object itself (e.g. `this.on('some-event', this.options.controller.someEventHandler);`), the object will hold on to references to the handlers, and prevent them from being GC'd.  Hopefuly, assuming one isn't doing any funkiness "higher up" in the logic, removing references to the object will sever the last path to the callbacks.  But, that's not an assumption that should be made, IMO: Backbone itself takes the stance that it's not a framework but a toolkit, and we all know how easy it is to find a hammer.
